BuildScript/BuildTasks/TestConfigurationsFilter.cs

```diff
@@ -0,0 +1,269 @@
+    private readonly IReadOnlyCollection<string> _enabledTargetRuntimes;
+    private readonly IReadOnlyCollection<string> _supportedConfigurationIDs;
+    private readonly IReadOnlyCollection<string> _enabledConfigurationIDs;
+    private readonly IReadOnlyCollection<string> _enabledExecutionRuntimes;
+    private readonly IReadOnlyCollection<string> _supportedTargetRuntimes;
+
+    public TestConfigurationsFilter (
+        IReadOnlyCollection<string> supportedBrowsers,
+        IReadOnlyCollection<string> enabledBrowsers,
+        IReadOnlyCollection<string> supportedDatabaseSystems,
+        IReadOnlyCollection<string> enabledDatabaseSystems,
+        IReadOnlyCollection<string> supportedPlatforms,
+        IReadOnlyCollection<string> enabledPlatforms,
+        IReadOnlyCollection<string> supportedExecutionRuntimes,
+        IReadOnlyCollection<string> enabledExecutionRuntimes,
+        IReadOnlyCollection<string> supportedTargetRuntimes,
+        IReadOnlyCollection<string> enabledTargetRuntimes,
+        IReadOnlyCollection<string> supportedConfigurationIDs,
+        IReadOnlyCollection<string> enabledConfigurationIDs)
+    {
+      _supportedBrowsers = supportedBrowsers;
+      _enabledBrowsers = enabledBrowsers;
+      _supportedDatabaseSystems = supportedDatabaseSystems;
+      _enabledDatabaseSystems = enabledDatabaseSystems;
+      _supportedPlatforms = supportedPlatforms;
+      _enabledPlatforms = enabledPlatforms;
+      _enabledExecutionRuntimes = enabledExecutionRuntimes;
+      _supportedTargetRuntimes = supportedTargetRuntimes;
+      _enabledTargetRuntimes = enabledTargetRuntimes;
+      _supportedConfigurationIDs = supportedConfigurationIDs;
+      _enabledConfigurationIDs = enabledConfigurationIDs;
+      _dockerImageToKeyDictionary = GetImageToKeyDictionary (supportedExecutionRuntimes);
+    }
+
+    public IReadOnlyCollection<TestConfiguration> GetFilteredTestConfigurations (IReadOnlyCollection<TestConfiguration> testConfigurationsToFilter)
+    {
+      EnsureNoEnabledMetadataIsEmpty();
+      EnsureAllMetadataAreSupported (testConfigurationsToFilter);
+
+      return testConfigurationsToFilter
+          .Where (HasEnabledBrowser)
+          .Where (HasEnabledPlatform)
+          .Where (HasEnabledDatabaseSystem)
+          .Where (HasEnabledExecutionRuntime)
+          .Where (HasEnabledTargetRuntime)
+          .Where (HasEnabledConfigurationID)
+          .ToArray();
+    }
+
+    private Dictionary<string, string> GetImageToKeyDictionary (IReadOnlyCollection<string> supportedExecutionRuntimes)
+    {
+      var splitKeyValuePairs = supportedExecutionRuntimes.Select (x => x.Split (new[] { "::" }, StringSplitOptions.RemoveEmptyEntries)).ToArray();
+      var values = splitKeyValuePairs.Select (x => x[1]).ToArray();
+      var duplicateValues = values.GroupBy (x => x).Where (x => x.Count() > 1).Select (x => x.Key).ToArray();
+
+      if (duplicateValues.Any())
```

Comment:
I'm reversing the dictionary and need to make sure the values are unique as well. By performing this check manually I can display a sensible error message instead of letting the exception bubble up

Reply:
Probably should be more verbose: I don't understand why it is not allowed to assign one docker image to multiple keys. What's the underlying assumption / business rule?